src/Microsoft.AspNetCore.Sockets.Common/TextMessageBatchFormatter.cs

```diff
@@ -0,0 +1,239 @@
+            cursor += length;
+
+            // Verify the length and trailer
+            if (cursor >= data.Length)
+            {
+                throw new FormatException("Unexpected end-of-message while reading Payload field.");
+            }
+            if (data[cursor] != ';')
+            {
+                throw new FormatException("Payload is missing trailer character ';'.");
+            }
+
+            // Read the data into a buffer
+            var buffer = new byte[length];
+            data.Slice(start, length).CopyTo(buffer);
+
+            // If the message is binary, we need to convert from Base64
+            if (messageType == MessageType.Binary)
+            {
+                // TODO: Use System.Binary.Base64 to handle this with less allocation
+
+                // Parse the data as Base64
+                var str = Encoding.UTF8.GetString(buffer);
+                buffer = Convert.FromBase64String(str);
+            }
+
+            return ReadableBuffer.Create(buffer).Preserve();
+        }
+
+        private static MessageType ParseType(ReadOnlySpan<byte> data, ref int cursor)
+        {
+            int start = cursor;
+
+            // Scan to a ':'
+            cursor = IndexOf((byte)':', data, cursor);
+            if (cursor >= data.Length)
+            {
+                throw new FormatException("Unexpected end-of-message while reading Type field.");
+            }
+
+            if (cursor - start != 1)
+            {
+                throw new FormatException("Type field must be exactly one byte long.");
+            }
+
+            switch (data[cursor - 1])
+            {
+                case (byte)'T': return MessageType.Text;
+                case (byte)'B': return MessageType.Binary;
+                case (byte)'C': return MessageType.Close;
+                case (byte)'E': return MessageType.Error;
+                default: throw new FormatException($"Unknown Type value: '{(char)data[cursor - 1]}'.");
+            }
+        }
+
+        private static int ParseLength(ReadOnlySpan<byte> data, ref int cursor)
+        {
+            int start = cursor;
+
+            // Scan to a ':'
+            cursor = IndexOf((byte)':', data, cursor);
+            if (cursor >= data.Length)
+            {
+                throw new FormatException("Unexpected end-of-message while reading Length field.");
+            }
+
+            // Parse the length
+            int length = 0;
+            for (int i = start; i < cursor; i++)
+            {
+                if (data[i] < '0' || data[i] > '9')
+                {
+                    throw new FormatException("Invalid length.");
+                }
+                length = (length * 10) + (data[i] - '0');
+            }
+
+            return length;
+        }
+
+        // There isn't a Span.IndexOf that takes a start point :(.
+        private static int IndexOf(byte c, ReadOnlySpan<byte> data, int start)
```

Comment:
You had previously indicated that Slice was still slow and that in general we should use indexing. Happy to use slice. It'll be cleaner and easier :)